src/crystal/system/unix/io_uring.cr

```diff
@@ -406,10 +405,15 @@ class Crystal::System::IoUring
       fatal_error "Failed to send submission entries to the kernel: #{Errno.new(-ret)}"
     end
 
+    # Requests are dropped if they are malformed. We never send those. This is a hard error because if we do,
+    # then some Fiber will never wake up and the completion event won't ever be sent.
     if @submission_queue.dropped > 0
       fatal_error "Submission queue has dropped entries: #{@submission_queue.dropped}"
     end
 
+    # This is a hard error because if some completion event was dropped, the waiting Fiber will leak.
+    # Linux 5.5+ have the NODROP feature and it will never fail to send a completion for a submitted event.
+    # For Linux before 5.45 we keep track of the maximum inflight requests and we don't submit above the limit.
```

Comment:
5.45? Should be 5.19-21 or something like that, I think?

Reply:
Oops, I meant "For Linux before 5.5".

I'm referring to this patch that added the NODROP feature. Before that, we have to keep track of the number of inflight requests to never risk overflowing the completion queue.

https://lore.kernel.org/io-uring/1fb145f2-7acb-ca76-b5e7-c17b2a30f31d@kernel.dk/T/

Reply:
Right. What I was referring to was https://lore.kernel.org/io-uring/20210416232837.EpkCt1h-_Fk1PTRiQA92ifC9l5pNLc2vk84GqYJcJJc@z/ , where they change that functionality again, so that it is fully up to the userspace to decide if they want to overcommit requests in flight.

Reply:
As per what I understood the new behavior is backward compatible in the sense that no completion event will ever be lost. With these recent changes, the kernel is able to queue an indefinite amount of backlog (consuming the process memory). Maybe we could apply some arbitrary limit of inflight requests on our side to prevent OOMK, but this seems not needed as each inflight request consumes one entire paused Fiber and their stacks consume way more memory than any number of backlogged completion events.